File: plugins/ios/chat-module/Sources/StarterChat/Views/MessageBubble.swift
import StarterAI
import SwiftUI

struct MessageBubble: View {
    let message: AIMessage

    var isUser: Bool { message.role == .user }

    var body: some View {
        HStack {
            if isUser { Spacer(minLength: 40) }
            Group {
                if message.content.isEmpty && message.role == .assistant {
                    TypingIndicator()
                } else {
                    Text(message.content)
                        .foregroundStyle(isUser ? Color.white : Color.primary)
                        .padding(.horizontal, 12)
                        .padding(.vertical, 8)
                }
            }
            .background(isUser ? Color.accentColor : Color(.secondarySystemBackground))
            .clipShape(RoundedRectangle(cornerRadius: 16))
            if !isUser { Spacer(minLength: 40) }
        }
    }
}

File: plugins/flutter/chat/lib/src/views/message_bubble.dart
import 'package:flutter/material.dart';
import 'package:starter_ai/starter_ai.dart';

import 'typing_indicator.dart';

/// Mirrors `MessageBubble.swift`.
class MessageBubble extends StatelessWidget {
  const MessageBubble({super.key, required this.message});

  final AIMessage message;

  bool get isUser => message.role == AIRole.user;

  @override
  Widget build(BuildContext context) {
    final theme = Theme.of(context);
    final isEmptyAssistant = message.content.isEmpty && message.role == AIRole.assistant;

    final bubble = Container(
      decoration: BoxDecoration(
        color: isUser ? theme.colorScheme.primary : theme.colorScheme.surfaceContainerHighest,
        borderRadius: BorderRadius.circular(16),
      ),
      padding: isEmptyAssistant
          ? EdgeInsets.zero
          : const EdgeInsets.symmetric(horizontal: 12, vertical: 8),
      child: isEmptyAssistant
          ? const TypingIndicator()
          : Text(
              message.content,
              style: TextStyle(
                color: isUser ? theme.colorScheme.onPrimary : theme.colorScheme.onSurface,
              ),
            ),
    );

    return Row(
      mainAxisAlignment: isUser ? MainAxisAlignment.end : MainAxisAlignment.start,
      children: [
        if (isUser) const SizedBox(width: 40),
        Flexible(child: bubble),
        if (!isUser) const SizedBox(width: 40),
      ],
    );
  }
}
